RecoParticleFlow/PFSimProducer/plugins/HGCalTrackCollectionProducer.cc

```diff
@@ -0,0 +1,275 @@
+  evt.getByToken(_src,trackHandle);
+  const auto& tracks = *trackHandle;  
+  
+  std::auto_ptr<reco::PFRecTrackCollection> outputInHGCal(new reco::PFRecTrackCollection);
+  std::auto_ptr<reco::PFRecTrackCollection> outputNotInHGCal(new reco::PFRecTrackCollection);
+
+  for ( unsigned int i = 0 ; i < tracks.size() ; i++) {
+    const auto track = tracks.ptrAt(i);
+    bool isGood = goodPtResolution(track->trackRef());
+    if (_debug) std::cout << "HGCalTrackCollectionProducer Track number " << i << " has a goodPtResolution result of " << isGood << std::endl;
+    if (!isGood) continue;
+    bool found = false;
+    const TrajectoryStateOnSurface myTSOS = trajectoryStateTransform::outerStateOnSurface(*(track->trackRef()), *(_tkGeom.product()),_bField.product());
+    auto detbegin = myTSOS.globalPosition().z() > 0 ? _plusSurface.begin() : _minusSurface.begin();
+    auto detend = myTSOS.globalPosition().z() > 0 ? _plusSurface.end() : _minusSurface.end();
+    for( auto det = detbegin; det != detend; ++det ) {  
+      if (_debug) std::cout << "at HGC detector: " << std::distance(detbegin,det) << std::endl;
+      unsigned layer_count = 1;
+      for( const auto& layer : *det ) {
+	if (_debug) std::cout << "  at DET layer: " << layer_count++ << std::endl;
+	TrajectoryStateOnSurface piStateAtSurface = _mat_prop->propagate(myTSOS, *layer);
+	if( piStateAtSurface.isValid() ) {
+	  if (_debug) std::cout << "Extrapolation is valid!" << std::endl;
+	  GlobalPoint pt = piStateAtSurface.globalPosition();
+	  if (pt.perp() < _diskOuterRadius) {
+	    if (pt.perp() > _diskInnerRadius) {
+	      if (_debug) std::cout << "(x,y,z,r)=(" << pt.x() << ", " << pt.y() << ", " << pt.z() << ", " << sqrt(pt.x()*pt.x() + pt.y()*pt.y()) << ")" << std::endl;
+	      if (_debug && fabs (track->trackRef()->eta()) < 1.47) std::cout << " ETA IN BARREL REGION: " << track->trackRef()->eta() 
+										  << " (PT: " << track->trackRef()->pt() << ")" << std::endl;
+	      found = true;
+	    } else {
+	      if (_debug) std::cout << " but r=" << pt.perp() << " < diskInnerRadius=" << _diskInnerRadius << " so skipping " << std::endl;
+	    }
+	  } else {
+	    if (_debug) std::cout << " but r=" << pt.perp() << " > diskOuterRadius=" << _diskOuterRadius << " so skipping " << std::endl;
+	  }
+	} else {
+	  if (_debug) std::cout << "Extrapolation is NOT valid!" << std::endl;
+	  //	  outputNotInHGCal->push_back(*track);
+	}
+      }
+    }
+    if (found) {
+      if (_debug) std::cout << " Track going to outputInHGCal pt eta " << track->trackRef()->pt() << " " << track->trackRef()->eta() << std::endl;
+      outputInHGCal->push_back(*track);
+    } else {
+      outputNotInHGCal->push_back(*track);
+    }
+  } // track loop
+
+  evt.put(outputInHGCal,"TracksInHGCal");
+  evt.put(outputNotInHGCal,"TracksNotInHGCal");
+}
+
+// Copied from https://github.com/cms-sw/cmssw/blob/CMSSW_8_1_X/RecoParticleFlow/PFProducer/plugins/importers/GeneralTracksImporter.cc#L149
```

Comment:
Why is this function copied? Couldn't it be made into a template friend function and used by both classes?